src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,443 @@
+
+	off_t result = ::lseek (fd, static_cast<off_t>(local_header_offset), SEEK_SET);
+	if (result < 0) {
+		log_fatal (LOG_DEFAULT, "Failed to seek to archive entry local header at offset %u", local_header_offset);
+		return false;
+	}
+
+	uint8_t local_header[ZIP_LOCAL_LEN];
+	uint8_t signature[4];
+
+	ssize_t nread = ::read (fd, local_header, static_cast<size_t>(ZIP_LOCAL_LEN));
+	if (nread < 0 || nread != ZIP_LOCAL_LEN) {
+		log_fatal (LOG_DEFAULT, "Failed to read local header at offset %u: ", local_header_offset, std::strerror (errno));
+		return false;
+	}
+
+	size_t index = 0;
+	if (!zip_read_field (local_header, ZIP_LOCAL_LEN, index, signature)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Local Header entry signature at offset %u", local_header_offset);
+		return false;
+	}
+
+	if (memcmp (signature, ZIP_LOCAL_MAGIC, sizeof(signature)) != 0) {
+		log_fatal (LOG_DEFAULT, "Invalid Local Header entry signature at offset %u", local_header_offset);
+		return false;
+	}
+
+	uint16_t file_name_length;
+	index = LH_FILE_NAME_LENGTH_OFFSET;
+	if (!zip_read_field (local_header, ZIP_LOCAL_LEN, index, file_name_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Local Header 'file name length' field at offset %u", (local_header_offset + index));
+		return false;
+	}
+
+	uint16_t extra_field_length;
+	index = LH_EXTRA_LENGTH_OFFSET;
+	if (!zip_read_field (local_header, ZIP_LOCAL_LEN, index, extra_field_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Local Header 'extra field length' field at offset %u", (local_header_offset + index));
+		return false;
+	}
+
+	data_start_offset = static_cast<uint32_t>(local_header_offset) + file_name_length + extra_field_length + ZIP_LOCAL_LEN;
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_extract_cd_info (uint8_t* buf, size_t buf_len, uint32_t& cd_offset, uint32_t& cd_size, uint16_t& cd_entries)
+{
+	static constexpr size_t EOCD_TOTAL_ENTRIES_OFFSET = 10;
+	static constexpr size_t EOCD_CD_SIZE_OFFSET       = 12;
+	static constexpr size_t EOCD_CD_START_OFFSET      = 16;
+
+	if (buf_len < ZIP_EOCD_LEN) {
+		log_fatal (LOG_DEFAULT, "Buffer to short for EOCD");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_TOTAL_ENTRIES_OFFSET, cd_entries)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'total number of entries' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_START_OFFSET, cd_offset)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory size' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_SIZE_OFFSET, cd_size)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory offset' field");
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_ensure_valid_params (uint8_t* buf, size_t buf_len, size_t index, size_t to_read)
+{
+	assert (buf != nullptr);
+	if (index + to_read > buf_len) {
+		log_fatal (LOG_DEFAULT, "Buffer too short to read %u bytes of data", to_read);
+		return false;
```

Comment:
I feel like this should be a "must pass" method, and if this check fails, the process should be exited.  As such, `zip_ensure_valid_params()` could return `void`: it either works, or the process dies.

If done, this should _not_ be done via assert, as it's too easy to disable asserts (are they even enabled on our release builds?), but should instead exit with a `FATAL_EXIT_*` value.

Reply:
I initially though it was a good idea, but then I realized that `zip_read_field` is a "generic" method to read data and if it fails with a fatal error message + exit, then all we'd see in the logcat would be a generic error message with no useful information to act on (e.g. what field wasn't read, from which part of ZIP was it read from etc) and while it could be solved by passing extra parameters to `zip_read_field` to print a more specific error message, that would require us using variable arguments and those would create a problem with the `log_*` functions which have no overload accepting `va_list` (and we do use multiple parameters when printing error messages here). All in all, I don't think it's worth it and that it would make for much better code.